### PR TITLE
Add device descriptors: sub_group_independent_forward_progress and preferred_vector_width

### LINKED ISSUE #286: Implement wrappers in C API and Python API for device descriptor information types
The list of device descriptor  [information types.](https://www.khronos.org/registry/SYCL/specs/sycl-2020/pdf/sycl-2020.pdf#%5B%7B%22num%22%3A133%2C%22gen%22%3A0%7D%2C%7B%22name%22%3A%22XYZ%22%7D%2C0%2C587.69%2Cnull%5D)

The example of implementations is here: https://github.com/IntelPython/dpctl/pull/147

- [x] info::device::device_type (info::device_type )
- [x] info::device::vendor_id (uint32_t)
- [x] info::device::max_compute_units (uint32_t )
- [x] info::device::max_work_item_dimensions (uint32_t)
- [x] info::device::max_work_item_sizes<1> (id<1>) - DPCTLDevice_GetMaxWorkItemSizes
- [x] info::device::max_work_item_sizes<2> (id<2>) - DPCTLDevice_GetMaxWorkItemSizes
- [x] info::device::max_work_item_sizes<3> (id<3>) - DPCTLDevice_GetMaxWorkItemSizes
- [x] info::device::max_work_group_size (size_t)
- [x] info::device::max_num_sub_groups (uint32_t)
- [x] info::device::sub_group_independent_forward_progress (bool) #314  
- [ ] info::device::sub_group_sizes (`std::vector<size_t>`)
- [x] info::device::preferred_vector_width_char, info::device::preferred_vector_width_short, info::device::preferred_vector_width_int, info::device::preferred_vector_width_long, info::device::preferred_vector_width_float, info::device::preferred_vector_width_double, info::device::preferred_vector_width_half (uint32_t) #314 
- [ ] info::device::native_vector_width_char, info::device::native_vector_width_short, info::device::native_vector_width_int, info::device::native_vector_width_long, info::device::native_vector_width_float, info::device::native_vector_width_double, info::device::native_vector_width_half (uint32_t)
- [ ] info::device::max_clock_frequency (uint32_t)
- [ ] info::device::address_bits (uint32_t)
- [ ] info::device::max_mem_alloc_size (uint64_t)
- [x] info::device::max_read_image_args (uint32_t) #318
- [x] info::device::max_write_image_args (uint32_t) #318
- [x] info::device::image2d_max_width (size_t) #315 
- [x] info::device::image2d_max_height (size_t) #315 
- [x] info::device::image3d_max_width (size_t) #315 
- [x] info::device::image3d_max_height (size_t) #315 
- [x] info::device::image3d_max_depth (size_t) #315 
- [ ] info::device::image_max_buffer_size (size_t)
- [ ] info::device::max_samplers (uint32_t)
- [ ] info::device::max_parameter_size (size_t)
- [ ] info::device::mem_base_addr_align (uint32_t)
- [ ] info::device::half_fp_config (`std::vector<info::fp_config>`)
- [ ] info::device::single_fp_config (`std::vector<info::fp_config>`)
- [ ] info::device::double_fp_config (`std::vector<info::fp_config>`)
- [ ] info::device::global_mem_cache_type (info::global_mem_cache_type)
- [ ] info::device::global_mem_cache_line_size (uint32_t)
- [ ] info::device::global_mem_cache_size (uint64_t)
- [ ] info::device::global_mem_size (uint64_t)
- [ ] info::device::local_mem_type (info::local_mem_­type)
- [ ] info::device::local_mem_size (uint64_t)
- [ ] info::device::error_correction_support (bool)
- [ ] info::device::atomic_memory_order_capabilities (`std::vector<memory_order>`)
- [ ] info::device::atomic_fence_order_capabilities (`std::vector<memory_order>`)
- [ ] info::device::atomic_memory_scope_capabilities (`std::vector<memory_scope>`)
- [ ] info::device::atomic_fence_scope_capabilities (`std::vector<memory_scope>`)
- [ ] info::device::profiling_timer_resolution (size_t)
- [ ] info::device::is_available (`bool`)
- [ ] info::device::execution_capabilities (`std::vector<info::execution_capability>`)
- [ ] info::device::built_in_kernel_ids (`std::vector<kernel_id>`)
- [ ] info::device::platform (`platform`)
- [x] info::device::name (`std::string`)
- [x] info::device::vendor (`std::string`)
- [x] info::device::driver_version (`std::string`)
- [ ] info::device::version (`std::string`)
- [ ] info::device::backend_version (`std::string`)
- [ ] info::device::aspects (`std::vector<aspect>`)
- [ ] info::device::parent_device (device)
- [ ] info::device::partition_max_sub_devices (uint32_t)
- [ ] info::device::partition_properties (`std::vector<info::partition_property>`)
- [ ] info::device::partition_affinity_domains (`std::vector<info::partition_affinity_domain>`)
- [ ] info::device::partition_type_property (`info::partition_property`)
- [ ] info::device::partition_type_affinity_domain (info::partition_affinity_domain)

Deprecated:
~~- [ ] info::device::image_support (bool)~~
~~- [ ] info::device::host_unified_memory (bool)~~
~~- [ ] info::device::is_endian_little (`bool`)~~
~~- [ ] info::device::is_compiler_available (`bool`)~~
~~- [ ] info::device::is_linker_available (`bool`)~~
~~- [ ] info::device::queue_profiling (`bool` )~~
~~- [ ] info::device::built_in_kernels (`std::vector<std::string>`)~~
~~- [ ] info::device::extensions (`std::vector<std::string>`)~~

Deprecated in SYCL 2020:
~~- [ ] info::device::max_constant_buffer_size (uint64_t)~~
~~- [ ] info::device::max_constant_args (uint32_t )~~
~~- [ ] info::device::profile (`std::string`)~~
~~- [ ] info::device::printf_buffer_size (size_t)~~
~~- [ ] info::device::preferred_interop_user_sync (bool)~~

Required #292

### LINKED ISSUE #314: Add device descriptors: sub_group_independent_forward_progress and preferred_vector_width
Part of the task #286 
 
 - [ ] info::device::sub_group_independent_forward_progress (bool)
 - [ ] info::device::preferred_vector_width_char, 
 - [ ] info::device::preferred_vector_width_short, 
 - [ ] info::device::preferred_vector_width_int, 
 - [ ] info::device::preferred_vector_width_long, 
 - [ ] info::device::preferred_vector_width_float, 
 - [ ] info::device::preferred_vector_width_double, 
 - [ ] info::device::preferred_vector_width_half (uint32_t)

### PR DESCRIPTION
**Pull request by 1e-to**
Add sub_group_independent_forward_progress and preferred_vector_width_char
Closes #314 , the part of #286

### CODE REVIEW

**Comment by 1e-to** [comment]
@diptorupd, @oleksandr-pavlyk could you please take a look?

**Comment by diptorupd** [comment]
@1e-to Looks good. @oleksandr-pavlyk once we have the types properly supported we should have a single `get_vector_width` call and pass in a dtype to lookup.

**Comment by oleksandr-pavlyk** [comment]
Can I suggest to not save properties in `SyclDevice` instance? The `get_info` call is not expensive, and Python property should just be making it. It is cleaner, I think.

**Comment by oleksandr-pavlyk** [comment]
For now this looks good. Feel free to merge. This will cause conflicts with `feature/queue_manager_v2` but I can resolve them.

**Comment by diptorupd** [comment]
> Can I suggest to not save properties in `SyclDevice` instance? The `get_info` call is not expensive, and Python property should just be making it. It is cleaner, I think.

I think it is better to make the change now before we add even more `get_info` wrappers. @1e-to please change these new attributes into `properties`. We can follow up with a different PR to change the existing attributes.

Also, rebase with master.

**Comment by diptorupd** [comment]
@1e-to can you rebase this with master and then change the attributes you added to SyclDevice class to properties.

File: dpctl-capi/tests/test_sycl_device_interface.cpp [hunk 1]
```diff
@@ -268,6 +269,128 @@ TEST_P(TestDPCTLSyclDeviceInterface, Chk_IsHost)
...
+TEST_P(TestDPCTLSyclDeviceInterface, Chk_GetPreferredVectorWidthDouble)
+{
+    DPCTLSyclDeviceRef DRef = nullptr;
+    size_t vector_width_double = 0;
+    EXPECT_NO_FATAL_FAILURE(DRef = DPCTLDevice_CreateFromSelector(DSRef));
+    if (!DRef)
+        GTEST_SKIP_("Device not found");
+    EXPECT_NO_FATAL_FAILURE(
+        vector_width_double = DPCTLDevice_GetPreferredVectorWidthDouble(DRef));
+    if (DPCTLDevice_HasAspect(DRef, DPCTL_SyclAspectToDPCTLAspectType(
```
Comment (diptorupd) [inline dpctl-capi/tests/test_sycl_device_interface.cpp:363]: Nice! You have been reading the spec a lot :)

**Review by diptorupd — APPROVED** [review]
Looks really good. Thanks!

**Review by oleksandr-pavlyk — APPROVED** [review]
Thanks @1e-to !

**Pull request merged**